Sources/Utility/ByteString.swift

```diff
@@ -0,0 +1,142 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// A `ByteString` represents a sequence of bytes.
+///
+/// This struct provides useful operations for working with buffers of
+/// bytes. Conceptually it is just a contiguous array of bytes (UInt8), but it
+/// contains methods and default behavor suitable for common operations done
+/// using bytes strings.
+///
+/// This struct *is not* intended to be used for significant mutation of byte
+/// strings, we wish to retain the flexibility to micro-optimize the memory
+/// allocation of the storage (for example, by inlining the storage for small
+/// strings or and by eliminating wasted space in growable arrays). For
+/// construction of byte arrays, clients should use the `OutputByteStream` class
+/// and then convert to a `ByteString` when complete.
+public struct ByteString: ArrayLiteralConvertible {
+    /// The buffer contents.
+    private var _bytes: [UInt8]
+
+    /// Create an empty byte string.
+    public init() {
+        _bytes = []
+    }
+    
+    /// Create a byte string from a byte array literal.
+    public init(arrayLiteral contents: UInt8...) {
+        _bytes = contents
+    }
+
+    /// Create a byte string from an array of bytes.
+    public init(_ contents: [UInt8]) {
+        _bytes = contents
+    }
+    
+    /// Create a byte string from an byte buffer.
+    public init(_ buffer: UnsafeBufferPointer<UInt8>) {
```

Comment:
`init<S : Sequence where S.Iterator.Element == UInt8>`.

We have an optimization that `Array(Sequence)` is a no-op if the Sequence is dynamically an array.